DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp

```diff
@@ -60,114 +60,177 @@ void Usage(string programName) {
+  if (err == true)
+    cout << "  FAIL! Please check " << path << "\n";
   else
-    return arr_next;
+    cout << "  PASSED!\n";
+  return err;
+}
+
+//
+// Compute heat on the device
+//
+void ComputeHeatDevice(float C, unsigned int num_p, unsigned int num_iter,
+		       float* arr_CPU) {
+  // Define the device queue
+  queue q;
+  cout << "Device\n";
+  cout << "  Kernel runs on " << q.get_device().get_info<info::device::name>() << "\n";
+
+  // Temperatures of the current and next iteration
+  float* arr_host = new float[num_p + 2];
+  float* arr_next_host = new float[num_p + 2];
+
+  Initialize(arr_host, arr_next_host, num_p + 2);
+
+  buffer arr_buf(arr_host, range(num_p + 2));
+  buffer arr_next_buf(arr_next_host, range(num_p + 2));
+
+  // Start timer
+  dpc_common::TimeInterval t_par;
+
+  // Iterate over timesteps
+  for (int i = 1; i <= num_iter; i++) {
+    q.submit([&](auto& h) {
+	       accessor arr(arr_buf, h);
+	       accessor arr_next(arr_next_buf, h);
+
+	       auto step =
+		 [=](id<1> k) {
+		   
+		   if (k == 0) {
+		   } else if (k == num_p + 1) {
+		     arr_next[k] = arr[k - 1];
+		   } else {
+		     arr_next[k] =
+		       C * (arr[k + 1] - 2 * arr[k] + arr[k - 1]) + arr[k];
+		   }
+		 };
+	       h.parallel_for(range{num_p + 2}, step);
+	     });    // end device queue
+    
+    // Swap buffers for next step
+    auto& swap = arr_buf;
+    arr_buf = arr_next_buf;
+    arr_next_buf = swap;
+  }
+
+  // Wait for all tasks to complete
+  q.wait_and_throw();
+
+  // Display time used to process all time steps
+  cout << "  Elapsed time: " << t_par.Elapsed() << " sec\n";
+
+  CompareResults("parallel", num_iter % 2 == 0 ? arr_host : arr_next_host, arr_CPU, num_p, C);
+
+  delete [] arr_host;
+  delete [] arr_next_host;
+}
+
+//
+// Compute heat on the device using USM
+//
+void ComputeHeatDeviceUSM(float C, unsigned int num_p, unsigned int num_iter,
+			  float* arr_CPU) {
+  // Use an in-order queue so the time steps are executed in-order
+  sycl::property_list properties{sycl::property::queue::in_order()};
+  queue q(properties);
+  cout << "Device with USM\n";
+  cout << "  Kernel runs on " << q.get_device().get_info<info::device::name>() << "\n";
+
+  // Temperatures of the current and next iteration
+  float* arr = malloc_shared<float>(num_p + 2, q);
+  float* arr_next = malloc_shared<float>(num_p + 2, q);
+
+  Initialize(arr, arr_next, num_p + 2);
+
+  // Start timer
+  dpc_common::TimeInterval t_par;
+
+  // Iterate over timesteps
+  for (int i = 1; i <= num_iter; i++) {
+    q.submit([&](auto& h) {
+	       auto step =
+		 [=](id<1> k) {
+		   
+		   if (k == 0) {
+		   } else if (k == num_p + 1) {
+		     arr_next[k] = arr[k - 1];
+		   } else {
+		     arr_next[k] =
+		       C * (arr[k + 1] - 2 * arr[k] + arr[k - 1]) +
+		       arr[k];
+		   }
+		 };
+	       h.parallel_for(range{num_p + 2}, step);
+	     });    // end device queue
+    
+    // Swap arrarys for next step
+    float* swap = arr;
```

Comment:
Use consistent code for swap.   Maybe switch to std::swap?   https://en.cppreference.com/w/cpp/algorithm/swap

Reply:
Using std::swap